### PR TITLE
Nigam backend

### PR DESCRIPTION
**Pull request by UtkarshNigamSkeps**
routefile added in routes folder and process file added in controllers

### CODE REVIEW

File: backend/src/routes/route.js [hunk 1]
```diff
@@ -1,3 +1,13 @@
-const router = new XPathExpression.router();
+const express=require('express')
+const upload=require('../middleware/multer')
+const processController= require('../controllers/process')
+const router=new express.Router();
 
-module.exports =  router;
\ No newline at end of file
+router.get('/', (req, res) => {
```
Comment (ntnmangla1) [inline backend/src/routes/route.js]: there should be some name for this get request

File: backend/src/routes/route.js [hunk 2]
```diff
@@ -1,3 +1,13 @@
...
+    res.render('index');
+});
+
+router.post('/process', upload.single('pdfFile'), processController.processPdf);ß
```
Comment (ntnmangla1) [inline backend/src/routes/route.js]: typo at the end of line

File: backend/src/routes/route.js [hunk 3]
```diff
@@ -1,3 +1,13 @@
-const router = new XPathExpression.router();
+const express=require('express')
+const upload=require('../middleware/multer')
+const processController= require('../controllers/process')
+const router=new express.Router();
 
-module.exports =  router;
\ No newline at end of file
+router.get('/', (req, res) => {
+    res.render('index');
```
Comment (ntnmangla1) [inline backend/src/routes/route.js]: render route should not be there in backend

File: backend/src/controllers/process.js [hunk 4]
```diff
@@ -1,5 +1,48 @@
...
 
+const runOcrCommand = require('../utility/ocr'); // Import your OCR function
+
+async function processPdf(req, res) {
+  const inputFile = req.file.path;
+  const outputFile = 'output.pdf'; // Replace with your output file path
+
+    await runOcrCommand(inputFile, outputFile).then(() => {
+    const pdfFilePath = 'output.pdf'; // Replace with your PDF file path
+
```
Comment (ntnmangla1) [inline backend/src/controllers/process.js]: format the code structure

File: backend/src/controllers/process.js [hunk 5]
```diff
@@ -1,5 +1,48 @@
-async function processPdf(){
+const fs = require('fs');
+const pdf = require('pdf-parse');
 
+const runOcrCommand = require('../utility/ocr'); // Import your OCR function
+
+async function processPdf(req, res) {
+  const inputFile = req.file.path;
```
Comment (ntnmangla1) [inline backend/src/controllers/process.js]: place this code in try and catch block

**Review by ntnmangla1 — CHANGES_REQUESTED** [review]

File: backend/src/controllers/process.js [hunk 6]
```diff
@@ -1,5 +1,48 @@
...
+        const startIndex = extractedText.indexOf(searchTerm);
+        const endIndex = startIndex + searchTerm.length;
+
+        const contextBefore = extractedText.substring(
+          Math.max(startIndex - 100, 0),
+          startIndex
+        );
+        const contextAfter = extractedText.substring(
+          endIndex,
+          Math.min(endIndex + 10, extractedText.length)
```
Comment (ntnmangla1) [inline backend/src/controllers/process.js]: why are we checking only for length=10?

File: backend/src/routes/route.js (hunk 2 above)
Comment (ntnmangla1) [inline backend/src/routes/route.js]: there should be validation in this request for body

**Review by ntnmangla1 — CHANGES_REQUESTED** [review]

File: backend/src/controllers/process.js (hunk 6 above)
Comment (UtkarshNigamSkeps) [reply backend/src/controllers/process.js]: we will change  this

File: backend/src/controllers/process.js (hunk 5 above)
Comment (UtkarshNigamSkeps) [reply backend/src/controllers/process.js]: done

File: backend/src/controllers/process.js (hunk 4 above)
Comment (UtkarshNigamSkeps) [reply backend/src/controllers/process.js]: done

File: backend/src/routes/route.js (hunk 1 above)
Comment (UtkarshNigamSkeps) [reply backend/src/routes/route.js]: removed

File: backend/src/routes/route.js (hunk 3 above)
Comment (UtkarshNigamSkeps) [reply backend/src/routes/route.js]: removed

File: backend/src/routes/route.js (hunk 2 above)
Comment (UtkarshNigamSkeps) [reply backend/src/routes/route.js]: removed

**Review by UtkarshNigamSkeps — COMMENTED** [review]
i have done all the suggested changes the sort ftn for length 10 and validation these we will add afterwards

File: backend/src/routes/route.js [hunk 7]
```diff
@@ -3,11 +3,11 @@ const upload=require('../middleware/multer')
 const processController= require('../controllers/process')
 const router=new express.Router();
 
-router.get('/', (req, res) => {
-    res.render('index');
-});
+// router.get('/', (req, res) => {
+//     res.render('index');
+// });
```
Comment (ntnmangla1) [inline backend/src/routes/route.js]: no need to comment it.. just remove it

File: backend/src/controllers/process.js [hunk 8]
```diff
@@ -4,43 +4,47 @@ const pdf = require('pdf-parse');
...
+        
+                res.send(`Search term found:${highlightedContent}`);
+              } else {
+                res.send('Search term not found.');
+              }
+            });
+          });
+    }
+    catch(e){
+        return res.send("e.message")
```
Comment (ntnmangla1) [inline backend/src/controllers/process.js]: it wont work... send =>.  res.error(e.message,"Process pdf failed");

File: backend/src/controllers/process.js [hunk 9]
```diff
@@ -4,43 +4,47 @@ const pdf = require('pdf-parse');
...
-        res.send('Search term not found.');
+   const inputFile = req.file.path;
+   const outputFile = 'output.pdf'; // Replace with your output file path
+    try{
+        await runOcrCommand(inputFile, outputFile).then(() => {
+            const pdfFilePath = 'output.pdf'; // Replace with your PDF file path
+        
+            // Read the PDF file
+            const pdfData = fs.readFileSync(pdfFilePath);
+
```
Comment (ntnmangla1) [inline backend/src/controllers/process.js]: formatting is not done... select the code and press right click and then select format the code

File: backend/src/controllers/process.js [hunk 10]
```diff
@@ -4,43 +4,47 @@ const pdf = require('pdf-parse');
...
+                const contextAfter = extractedText.substring(
+                  endIndex,
+                  Math.min(endIndex + 10, extractedText.length)
+                );
+        
+                const highlightedContent = `${searchTerm}${contextAfter}`;
+        
+                res.send(`Search term found:${highlightedContent}`);
+              } else {
+                res.send('Search term not found.');
```
Comment (ntnmangla1) [inline backend/src/controllers/process.js]: use res.success()

**Review by ntnmangla1 — CHANGES_REQUESTED** [review]

File: backend/src/controllers/process.js (hunk 8 above)
Comment (UtkarshNigamSkeps) [reply backend/src/controllers/process.js]: done

File: backend/src/controllers/process.js (hunk 9 above)
Comment (UtkarshNigamSkeps) [reply backend/src/controllers/process.js]: done

File: backend/src/routes/route.js (hunk 7 above)
Comment (UtkarshNigamSkeps) [reply backend/src/routes/route.js]: done

File: backend/src/controllers/process.js (hunk 10 above)
Comment (UtkarshNigamSkeps) [reply backend/src/controllers/process.js]: kk sir we will use this afterwards

**Review by ntnmangla1 — APPROVED** [review]
lftm

File: backend/src/controllers/process.js [hunk 11]
```diff
@@ -1,5 +1,50 @@
...
+const fs = require('fs');
+const pdf = require('pdf-parse');
 
+const runOcrCommand = require('../utility/ocr'); // Import your OCR function
+
+async function processPdf(req, res) {
+    const inputFile = req.file.path;
+    const outputFile = 'output.pdf'; // Replace with your output file path
+    try {
+        await runOcrCommand(inputFile, outputFile).then(() => {
```
Comment (pradeep-sRawat) [inline backend/src/controllers/process.js]: why use await and .then together ?

**Review by pradeep-sRawat — CHANGES_REQUESTED** [review]

File: backend/src/controllers/process.js (hunk 11 above)
Comment (UtkarshNigamSkeps) [reply backend/src/controllers/process.js]: await removed

File: backend/src/controllers/process.js [hunk 12]
```diff
@@ -1,5 +1,50 @@
-async function processPdf(){
+const fs = require('fs');
+const pdf = require('pdf-parse');
 
+const runOcrCommand = require('../utility/ocr'); // Import your OCR function
+
+async function processPdf(req, res) {
+    const inputFile = req.file.path;
+    const outputFile = 'output.pdf'; // Replace with your output file path
```
Comment (s-taran00) [inline backend/src/controllers/process.js:8]: Get this path from req as well

File: backend/src/controllers/process.js [hunk 13]
```diff
@@ -1,5 +1,50 @@
...
+    try {
+        runOcrCommand(inputFile, outputFile).then(() => {
```
Comment (s-taran00) [inline backend/src/controllers/process.js:10]: use await instead of callback

File: backend/src/controllers/process.js [hunk 14]
```diff
@@ -1,5 +1,50 @@
...
+            const pdfFilePath = 'output.pdf'; // Replace with your PDF file path
```
Comment (s-taran00) [inline backend/src/controllers/process.js:11]: +1

File: backend/src/controllers/process.js [hunk 15]
```diff
@@ -1,5 +1,50 @@
...
+
+            // Read the PDF file
+            const pdfData = fs.readFileSync(pdfFilePath);
+
+            // Parse the PDF data
+            pdf(pdfData).then(data => {
```
Comment (s-taran00) [inline backend/src/controllers/process.js:17]: use await instead of callback

File: backend/src/controllers/process.js [hunk 16]
```diff
@@ -1,5 +1,50 @@
...
+
+                // Now you can search for specific data in the extracted text
+                const searchTerm = req.body.searchTerm; // Replace with the data you want to find
+
+                if (extractedText.includes(searchTerm)) {
+                    const startIndex = extractedText.indexOf(searchTerm);
+                    const endIndex = startIndex + searchTerm.length;
+
+                    const contextBefore = extractedText.substring(
+                        Math.max(startIndex - 100, 0),
```
Comment (s-taran00) [inline backend/src/controllers/process.js:28]: why is this hardcoded 100?

File: backend/src/controllers/process.js [hunk 17]
```diff
@@ -1,5 +1,50 @@
...
+                        startIndex
+                    );
+                    const contextAfter = extractedText.substring(
+                        endIndex,
+                        Math.min(endIndex + 10, extractedText.length)
```
Comment (s-taran00) [inline backend/src/controllers/process.js:33]: Why is this hardcoded 10?

**Review by pradeep-sRawat — APPROVED** [review]

File: backend/src/controllers/process.js [hunk 18]
```diff
@@ -1,5 +1,50 @@
...
+                    );
+
+                    const highlightedContent = `${searchTerm}${contextAfter}`;
```
Comment (s-taran00) [inline backend/src/controllers/process.js:36]: This is never used

**Pull request merged**